#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <QMouseEvent>
#include "utils/tools.h"
#include "constants.h"
#include <vector>
#include <dirent.h>
#include "UI/picturecut.h"
#include <windows.h>
#include <QFileDialog>
#include "socket/filesock.h"
using namespace std;

MainWindow::MainWindow(QWidget *parent) : QWidget(parent),
                                          client(DataSock::get()),
                                          ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    ui->biaoqingFrame->setVisible(false);
    tmpCmd = QString();
    //å»çª—å£è¾¹æ¡†
    setWindowFlags(Qt::FramelessWindowHint | windowFlags());
    // TODOï¼šæ²¡æœ‰çª—å£è¾¹æ¡†æ€ä¹ˆé¡¶éƒ¨æ‹–åŠ¨ï¼Ÿç»§æ‰¿QLabelï¼Œé‡å†™mouseMoveEventå’Œpress...ï¼Œé¡¶éƒ¨ä¸¤ä¸ªlabelå°±å¯ä»¥å®ç°æ‹–åŠ¨çª—å£äº†
}

void MainWindow::clearAllMessage()
{
    while (ui->listWidget->count() > 0)
    {
        delete ui->listWidget->item(ui->listWidget->count() - 1);
    }
}

MainWindow::MainWindow(QString ud, QWidget *parent) : QWidget(parent),
                                                      udata(ud),
                                                      client(DataSock::get()),
                                                      ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    ui->biaoqingFrame->setVisible(false);
    ui->indowLabel->setText(ud);
    friendList = new Ui::FriendList(ui->friendList);
    initConnection();
    _initHandler();
    //    imag = new WNCimage(ud, client);
    //å»çª—å£è¾¹æ¡†
    setWindowFlags(Qt::FramelessWindowHint | windowFlags());
    // TODOï¼šæ²¡æœ‰çª—å£è¾¹æ¡†æ€ä¹ˆé¡¶éƒ¨æ‹–åŠ¨ï¼Ÿç»§æ‰¿QLabelï¼Œé‡å†™mouseMoveEventå’Œpress...ï¼Œé¡¶éƒ¨ä¸¤ä¸ªlabelå°±å¯ä»¥å®ç°æ‹–åŠ¨çª—å£äº†
    //    connect(client, SIGNAL(readyRead()), this, SLOT(hadreadyread()));
    qDebug() << "ud=" << ud;
    //å‘é€å¥½å‹åˆ—è¡¨è¯·æ±‚
    std::string data = Encoder_askfriendsList(udata.toStdString());
    QString packData = QString::fromStdString(data);
    client.sendMessage(packData);
}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::_initHandler()
{

    client.addCallback("askfriendsList", [=](const Json::Value &jtmp)
                       {
        std::string username, info;
        bool state;
        std::string userimage;
        vector<pair<string, string>> uinfolist;
        if (Decoder_askfriendsList(jtmp, username, uinfolist, userimage) == 0)
        {
            qDebug("askfriends data back from server error/n");
            return;
        }
        createRequireTask(QString::fromStdString(userimage), "./assets/"+ QString::fromStdString(username), [=](FileSock *fsk, const QFileInfo &fileName){
            if(fileName.absolutePath().length()!=0)
                user_image = QImage(fileName.absoluteFilePath());
            changeMyIcon(&user_image);
        }, "head");
        std::vector<Ui::User> userList;
        for(pair<string, string> frdinfo: uinfolist)
        {
            Ui::User frd;
            frd.userName = frdinfo.first;
            frd.image = QImage();
            userList.push_back(frd);
            msgRcd[QString::fromStdString(frdinfo.first)]=MessageRecord();
        }

        friendList->initFriendList(userList);
        for(pair<string, string> frdinfo: uinfolist){

            createRequireTask(QString::fromStdString(frdinfo.second), "./assets/"+QString::fromStdString(username)+"/friendheads",
                              [=](FileSock *fsk, const QFileInfo &fileInfo){
                QImage icon;
                if(fileInfo.absolutePath().length() != 0)
                    icon.load(fileInfo.absoluteFilePath());
                changeIcon(QString::fromStdString(frdinfo.first), icon);
            }, QString::fromStdString(frdinfo.first) + "head");
        }
    });
    client.addCallback("add_group_chat", [=](const Json::Value &jtmp)
    {
        int groupid;
        std::vector<std::string> username;
        bool state;
        string groupname;
        if(Decoder_add_group_chat(jtmp, username,groupid,state, groupname) == 0)
        {
            qDebug("group_chat data back from server error/n");
            return;
        }
        Ui::Group group;
        group.groupid = groupid;
        group.member = username;
        friendList->addGroupInfo(group);
    });
    client.addCallback("chat", [=](const Json::Value &jtmp)
                       {
        std::string sender_username, msg, time;
        if (Decoder_chat(jtmp, sender_username, msg, time) == 0)
        {
            qDebug("chat data back from server error/n");
            return;
        }
        pushMessageIntoChatWindow(false, QString::fromStdString(msg), QString::number(QDateTime::currentDateTime().toTime_t()), &chattingInfo.chatFriend[0].image);
        msgRcd[QString::fromStdString(sender_username)].push_back(MessageInfo(QString::fromStdString(sender_username), QString::fromStdString(msg), QString::fromStdString(time), MessageInfo::PLAIN_TXT));
        MoveFps(); });
    client.addCallback("chat_group", [=](const Json::Value &jtmp)
                       {
        int groupid;
        std::string username, msg, time;
        if (Decoder_chat(jtmp, username, msg, time) == 0)
        {
            qDebug("chat group data back from server error/n");
            return;
        }
        Ui::User sender_;
        for(Ui::User user: chattingInfo.chatFriend)
            if(user.userName==username)
                sender_ = user;
        pushMessageIntoChatWindow(false, QString::fromStdString(msg), QString::fromStdString(time),&sender_.image);
        MoveFps(); });
    client.addCallback("addfriends", [=](const Json::Value &jtmp)
                       {
        // mainwindowé‡Œå¤„ç†åŠ å¥½å‹çš„è¿”å›ä¿¡æ¯
        std::string friend_username;
        bool state;
        std::string friendimage;
        if (Decoder_addfriends(jtmp, friend_username, state, friendimage) == 0)
        {
            qDebug("addfriends data back from server error/n");
            return;
        }
        if (state == 1)
        {
            /*Todo:åœ¨åˆ—è¡¨é‡Œæ·»åŠ è¿™ä¸ªæ–°çš„å¥½å‹*/
            createRequireTask(QString::fromStdString(friendimage), "./assets/"+udata+"/friendheads",[=](FileSock *fsk, const QFileInfo &fileName){
                QImage friend_image;
                if(fileName.absolutePath().length()!=0)
                    friend_image.load(fileName.absoluteFilePath());
                Ui::User f;
                f.userName = friend_username;
                f.image = friend_image;
                friendList->insertToGroup("default", vector<Ui::User>{f});
                msgRcd[QString::fromStdString(friend_username)]=MessageRecord();
                QMessageBox::information(this, "æç¤º", ("æˆåŠŸæ·»åŠ å¥½å‹" + friend_username).c_str()); //
            }, QString::fromStdString(friend_username+"head"));

        }
        else
            QMessageBox::information(this, "æç¤º", "è¯¥å¥½å‹ä¸å­˜åœ¨"); });
    client.addCallback("chatfile", [=](const Json::Value &jtmp)
                       {
        std::string sender_username, fn, time;
        if (Decoder_chatfile(jtmp, fn, sender_username, time) == 0)
        {
            qDebug("chatfile data back from server error/n");
            return;
        }
        createRequireTask(QString::fromStdString(fn), "./assets/"+udata+"/FileRecv", [=](FileSock *sk, const QFileInfo &filesName){
            QString filename;
            QString cat;
            if(filesName.absolutePath().length()==0)
            {
                filename = "";
                qDebug()<<"fnm "<<"";
                cat="";
            }
            else
            {
                filename = filesName.absoluteFilePath();
                qDebug()<<"fnm "<<filename;
                cat = filename.right(filename.size() - filename.lastIndexOf(".") - 1);
            }

            qDebug() << "æ–‡ä»¶ç±»å‹ï¼š" << cat;
            if (cat == "jpg" || cat == "jpeg" || cat == "svg" || cat == "bmp" || cat == "png" || cat == "gif")
            {
                QImage chat_image = QImage(filename);
                //å¤„ç†è¿”å›æ¥çš„chat_imageå›¾ç‰‡ä¿¡æ¯
                pushImageIntoChatWindow(false, chat_image, QString::fromStdString(time), &chattingInfo.chatFriend[0].image);
                msgRcd[QString::fromStdString(sender_username)].push_back(MessageInfo(QString::fromStdString(sender_username), chat_image, QString::fromStdString(time), MessageInfo::IMAGE));
                MoveFps();
            }
            else
            {
                pushFileIntoChatWindow(false, filename, QString::fromStdString(time), &chattingInfo.chatFriend[0].image);
                msgRcd[QString::fromStdString(sender_username)].push_back(MessageInfo(QString::fromStdString(sender_username), filename, QString::fromStdString(time), MessageInfo::FILE));
                MoveFps();
            }
        }); });
    client.addCallback("msgrecord", [=](const Json::Value &jtmp)
                       {
        string sdr = jtmp["sender"].asString();
        string rcv = jtmp["receiver"].asString();
        int isy = jtmp["isYou"].asInt();
        string msg = jtmp["msg"].asString();
        string time = jtmp["time"].asString();
        if(isy)
        {
            msgRcd[QString::fromStdString(rcv)].push_back(MessageInfo(QString::fromStdString(sdr), QString::fromStdString(msg), QString::fromStdString(time), MessageInfo::PLAIN_TXT));   
        }
        else
        {
            msgRcd[QString::fromStdString(sdr)].push_back(MessageInfo(QString::fromStdString(sdr), QString::fromStdString(msg), QString::fromStdString(time), MessageInfo::PLAIN_TXT));
        } });
    client.addCallback("submit_image", [=](const Json::Value &jtmp){
        string username = jtmp["username"].asString();
        string imgPath = jtmp["image"].asString();
//        createRequireTask(QString::fromStdString(imgPath), "./assets/friend")
    });
}

void MainWindow::on_hideButton_clicked()
{
    showMinimized();
}

void MainWindow::on_closeButton_clicked()
{
    close();
}

void MainWindow::on_send_clicked()
{
    QString msg = ui->textEdit->toPlainText();
    if (msg.isEmpty())
        return; //ç©ºå­—ç¬¦ä¸å‘é€

    QString time;
    pushMessageIntoChatWindow(true, msg, time = QString::number(QDateTime::currentDateTime().toTime_t()), &user_image, false);
    std::vector<std::string> usersList;
    for(auto v:chattingInfo.chatFriend)
        usersList.push_back(v.userName);
    QString packData;
    if(!isgroup)
    {
        msgRcd[QString::fromStdString(chattingInfo.chatFriend[0].userName)].
            push_back(MessageInfo(QString::fromStdString(chattingInfo.chatFriend[0].userName),msg,time,MessageInfo::PLAIN_TXT));
    //å‘é€æ•°æ®åè®®

        std::string data = Encoder_chat(udata.toStdString(), msg.toStdString(), time.toStdString(), usersList);
        packData = QString::fromStdString(data);
    }
    else
    {
        std::string data = Encoder_chat_group(udata.toStdString(), groupid, msg.toStdString(), time.toStdString());
        packData = QString::fromStdString(data);
    }
    client.sendMessage(packData);
    ui->textEdit->clear();
}

/*
 * pushMessageIntoChatWindowå‡½æ•°è¯´æ˜
 * bool typeï¼šå‘ç»™åˆ«äººæ¶ˆæ¯ï¼Œå³ä¾§æ°”æ³¡==trueï¼›æ¥æ”¶åˆ°æ¶ˆæ¯ï¼Œå·¦ä¾§æ°”æ³¡==falseã€‚
 * QString msgï¼šå‘é€/æ¥æ”¶çš„æ–‡æœ¬ã€‚
 * QString timeï¼šæ¶ˆæ¯æ—¶é—´ã€‚
 *      å‚è€ƒï¼štime = QString::number(QDateTime::currentDateTime().toTime_t());
 * bool isSending = falseï¼šæ˜¯å¦æ­£åœ¨å‘é€ï¼Œå¤§æ¦‚ç‡ç”¨ä¸åˆ°æ‰€ä»¥å…ˆè®¾ç½®æˆfalse
 *      ä»…å‘é€æ¶ˆæ¯æœ‰æ•ˆï¼Œæ¥æ”¶æ¶ˆæ¯æ—¶è¿™é‡Œå¯ä»¥éšä¾¿èµ‹å€¼ã€‚
 * QImage* image = NULL:å¤´åƒæŒ‡é’ˆï¼Œé»˜è®¤ä¸ºNULLï¼Œåˆ™æ˜¾ç¤ºé»˜è®¤å¤´åƒã€‚
 *
 * *******æ³¨æ„********
 * â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“ç‰¹åˆ«æ³¨æ„â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“
 * â‘    UIä¿®æ”¹æ—¶éœ€è¦å…¨å±€æœç´¢æ›¿æ¢widthWithoutListWidgetçš„å€¼
 *      è¯¥å€¼æ˜¯æ•´ä¸ªmainWindowå»æ‰ListWidgetå¯¹è¯æ¡†çš„å®½åº¦
 *      ä¸æ›¿æ¢å¯èƒ½å¯¹è¯æ¡†æ˜¾ç¤ºå¤§å°å¼‚å¸¸
 * â‘¡   TODOï¼šæ£€æŸ¥å¯èƒ½æœ‰å†…å­˜æº¢å‡ºï¼Œæ³¨æ„delete ui->listWidget->item(i);
 *      forå¾ªç¯å°†æ‰€æœ‰éƒ½åˆ æ‰
 * â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘
 * æ³¨æ„å¤„ç†ç©ºæ¶ˆæ¯ï¼Œç©ºæ¶ˆæ¯æœ€å¥½åˆ«ä¼ è¿›æ¥ï¼Œä¼ è¿›æ¥ä¹Ÿä¸ä¼šæ˜¾ç¤º
 * å¦‚æœæ˜¯å‘æ¶ˆæ¯ï¼Œæ³¨æ„æ¸…ç©ºè¾“å…¥æ¡†
 * TODOï¼šä¸æ¸…æ¥šåˆ†è¾¨ç‡å’Œç¼©æ”¾çš„å½±å“ï¼Œå¯èƒ½ä¼šæœ‰å°æ¦‚ç‡é•¿æ–‡æœ¬æ˜¾ç¤ºæ¨ªå‘æ‹–åŠ¨æ¡
 */
void MainWindow::pushMessageIntoChatWindow(bool type, QString msg, QString time, QImage *image, bool isSending)
{

    if (msg == "")
    {
        QMessageBox::warning(this, "è­¦å‘Š", "å‘é€å†…å®¹ä¸èƒ½ä¸ºç©º", QMessageBox::Ok, QMessageBox::Ok);
        return;
    }
    qDebug() << "addMessage" << msg << time << ui->listWidget->count();
    if (type)
    {
        qDebug() << "self";
        if (isSending)
        {
            qDebug() << "sending";
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealMessage(messageW, item, msg, time, ChatMessageWidget::User_Me, image);
        }
        else
        {
            qDebug() << "notSendig";
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealMessage(messageW, item, msg, time, ChatMessageWidget::User_Me, image);
            messageW->setTextSuccess();
        }
    }
    else
    {
        qDebug()<<"hahahahaa";
        dealMessageTime(time);
        qDebug()<<"hahahahaa";
        ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
        qDebug()<<"hahahahaa";
        QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
        qDebug()<<"hahahahaa";
        dealMessage(messageW, item, msg, time, ChatMessageWidget::User_She, image);
        qDebug()<<"hahahahaa";
    }
    ui->listWidget->setCurrentRow(ui->listWidget->count() - 1);
}

void MainWindow::dealMessage(ChatMessageWidget *messageW, QListWidgetItem *item, QString text, QString time, ChatMessageWidget::User_Type type, QImage *image)
{
    messageW->setFixedWidth(ui->listWidget->width() - 25);
    QSize size = messageW->fontRect(text);
    item->setSizeHint(size);
    messageW->setText(text, time, size, type, image);
    ui->listWidget->setItemWidget(item, messageW);
}

void MainWindow::dealMessageTime(QString curMsgTime)
{
    bool isShowTime = false;
    if (ui->listWidget->count() > 0)
    {
        QListWidgetItem *lastItem = ui->listWidget->item(ui->listWidget->count() - 1);
        ChatMessageWidget *messageW = (ChatMessageWidget *)ui->listWidget->itemWidget(lastItem);
        int lastTime = messageW->time().toInt();
        int curTime = curMsgTime.toInt();
        qDebug() << "curTime lastTime:" << curTime - lastTime;
        isShowTime = ((curTime - lastTime) > 60); // ä¸¤ä¸ªæ¶ˆæ¯ç›¸å·®ä¸€åˆ†é’Ÿ
                                                  //        isShowTime = true;
    }
    else
    {
        isShowTime = true;
    }
    if (isShowTime)
    {
        ChatMessageWidget *messageTime = new ChatMessageWidget(ui->listWidget->parentWidget());
        QListWidgetItem *itemTime = new QListWidgetItem(ui->listWidget);
        QSize size = QSize(ui->listWidget->width() - 25, 40);
        messageTime->resize(size);
        itemTime->setSizeHint(size);
        qDebug() << curMsgTime;
        messageTime->setText(curMsgTime, curMsgTime, size, ChatMessageWidget::User_Time);
        ui->listWidget->setItemWidget(itemTime, messageTime);
    }
}

void MainWindow::on_pushButton_addfriend_clicked()
{
    add = new searchFriends(udata);
    add->show();
}



void MainWindow::on_pushButton_image_clicked()
{
    QString image_addr = QFileDialog::getOpenFileName(this, tr("æ‰“å¼€å›¾ç‰‡"), "/", "Image Files(*.jpg *.jpeg *.bmp *.png)");
    qDebug()<<image_addr;
    if (image_addr.length() == 0)
    {
        return;
    }

    PictureCut *w = new PictureCut(image_addr.toStdString());
    w->show();
    connect(w, SIGNAL(getSelectedPicture(Ui::headImage)), this, SLOT(submitheadImage(Ui::headImage)));
}

void MainWindow::submitheadImage(Ui::headImage hdimage)
{
    QString image_addr = QString::fromStdString(hdimage.path);
    createSendTask(
        udata, image_addr, [=](FileSock *, const QFileInfo &fileName, const QString &serverFileName)
        {
        std::string data = Encoder_submit_image(udata.toStdString(), serverFileName.toStdString());
        QString packData = QString::fromStdString(data);
        client.sendMessage(packData);
        qDebug()<<fileName.absoluteFilePath();
        user_image = QImage(fileName.absoluteFilePath());
        changeMyIcon(&user_image); },
        "head");
}

void MainWindow::on_pushButton_send_image_clicked()
{
    QString image_addr = QFileDialog::getOpenFileName(this);
    if (image_addr.length() == 0)
    {
        return;
    }
    std::vector<std::string> userslist; // userListéœ€è¦ä»å¥½å‹æ å¯¼å…¥çš„ï¼Œè¿™é‡Œå›ºå®šçš„å ä¸ªä½ç½®
    for(auto v:chattingInfo.chatFriend)
        userslist.push_back(v.userName);
    sendChatImage(image_addr, userslist, [=]()
    {});
}
void MainWindow::on_biaoqingButton_clicked()
{
    if (!(ui->biaoqingFrame->visibleRegion().isNull()))
    {
        ui->biaoqingFrame->setVisible(false);
    }
    ui->biaoqingFrame->setVisible(true);
}
void MainWindow::on_bButton1_clicked()
{
    sendChatImage(":/assets/å¾®ç¬‘.jpg", vector<string>{chattingInfo.chatFriend[0].userName}, [=]()
                  { ui->biaoqingFrame->setVisible(false); });
}
void MainWindow::on_bButton2_clicked()
{
    sendChatImage(":/assets/å¥½ï¼.jpg", vector<string>{chattingInfo.chatFriend[0].userName}, [=]()
                  { ui->biaoqingFrame->setVisible(false); });
}
void MainWindow::on_bButton3_clicked()
{
    sendChatImage(":/assets/å¤§å“­ï¼.jpg", vector<string>{chattingInfo.chatFriend[0].userName}, [=]()
                  { ui->biaoqingFrame->setVisible(false); });
}
void MainWindow::on_bButton4_clicked()
{
    sendChatImage(":/assets/å†·ç¬‘ï¼.jpg", vector<string>{chattingInfo.chatFriend[0].userName}, [=]()
                  { ui->biaoqingFrame->setVisible(false); });
}
void MainWindow::on_bButton5_clicked()
{
    sendChatImage(":/assets/è£…å‚».jpeg", vector<string>{chattingInfo.chatFriend[0].userName}, [=]()
                  { ui->biaoqingFrame->setVisible(false); });
}
void MainWindow::on_bButton6_clicked()
{
    sendChatImage(":/assets/ç‚¹èµ.jpeg", vector<string>{chattingInfo.chatFriend[0].userName}, [=]()
                  { ui->biaoqingFrame->setVisible(false); });
}
void MainWindow::on_bButton7_clicked()
{
    sendChatImage(":/assets/æ¸´æœ›.jpeg", vector<string>{chattingInfo.chatFriend[0].userName}, [=]()
                  { ui->biaoqingFrame->setVisible(false); });
}
void MainWindow::on_bButton8_clicked()
{
    sendChatImage(":/assets/æ±‚æ±‚ä½ äº†.jpeg", vector<string>{chattingInfo.chatFriend[0].userName}, [=]()
                  { ui->biaoqingFrame->setVisible(false); });
}
void MainWindow::on_bButton9_clicked()
{
    sendChatImage(":/assets/aaa.png", vector<string>{chattingInfo.chatFriend[0].userName}, [=]()
                  { ui->biaoqingFrame->setVisible(false); });
}

void MainWindow::initConnection()
{
    connect(friendList, SIGNAL(openChatroom(QVariant,int)), this, SLOT(startChatting(QVariant,int)));
    connect(friendList, SIGNAL(delFriend(QVariant)), this, SLOT(deleteFriend(QVariant)));
}
void MainWindow::startChatting(QVariant variant,int a)
{
    ui->cover_label->setVisible(false);
    qDebug()<<"jinlaile";
    clearAllMessage();
    qDebug()<<a;
    isgroup=a;
    if(!a)
    {
        Ui::User chatFriend = variant.value<Ui::User>();
        qDebug() << "here";
        qDebug() << QString(chatFriend.userName.c_str());
        ui->chatName->setText("   " + QString(chatFriend.userName.c_str()));
        chattingInfo.chatFriend.clear();
        chattingInfo.chatFriend.push_back(chatFriend);

        chattingInfo.record = &msgRcd[QString::fromStdString(chatFriend.userName)];
        qDebug() << msgRcd[QString::fromStdString(chatFriend.userName)].size();
        for (MessageInfo rec : *chattingInfo.record)
        {
            bool isyou = (rec.Sender == udata);
            QImage *img = isyou?&user_image:&chattingInfo.chatFriend[0].image;
            if(rec.type == MessageInfo::PLAIN_TXT)
            {
                pushMessageIntoChatWindow(isyou, rec.msg, rec.time, img);
            }
            else if(rec.type == MessageInfo::IMAGE)
            {
                pushImageIntoChatWindow(isyou, rec.img, rec.time, img);
            }
            else if(rec.type == MessageInfo::FILE)
            {
                pushFileIntoChatWindow(isyou, rec.filePath, rec.time, img);
            }
        }
    }
    else
    {
        Ui::Group group = variant.value<Ui::Group>();
        qDebug() << "here";
        qDebug() << QString::number(group.groupid);
        qDebug()<<"hahaha";
        groupid = group.groupid;
        ui->chatName->setText("   " + QString::number(group.groupid));
        chattingInfo.chatFriend.clear();
        for(std::string name:group.member)
        {
            if(QString::fromStdString(name) == udata) continue;
            qDebug()<<QString::fromStdString(name)+" hahaha";
            QTreeWidgetItem* item = friendList->friends[name];
            qDebug()<<QString::fromStdString(name)+" hahaha";
            Ui::User chatFriend = item->data(0,friendList->UserInfo).value<Ui::User>();

            chattingInfo.chatFriend.push_back(chatFriend);
        }

    }

}

void MainWindow::deleteFriend(QVariant variant)
{
    qDebug() << "in delete friend function";
    Ui::User friendToDel = variant.value<Ui::User>();
    qDebug() << "del friend: " << friendToDel.userName.c_str();
    // è¿”å›æœåŠ¡å™¨
    // friendToDel.userName: åˆ é™¤çš„å¥½å‹çš„åå­—
    std::string data = Encoder_deletefriends(udata.toStdString(), friendToDel.userName);
    QString packData = QString::fromStdString(data);
    client.sendMessage(packData);
    if (chattingInfo.chatFriend[0].userName == friendToDel.userName)
    {
        ui->cover_label->setVisible(true);
    }
}
//ä»¥ä¸‹æ˜¯æ‹–åŠ¨çª—å£çš„ä»£ç 
void MainWindow::mousePressEvent(QMouseEvent *ev)
{
    if (ev->button() == Qt::LeftButton)
    {
        //æ±‚åæ ‡å·®å€¼
        //å½“å‰ç‚¹å‡»åæ ‡-çª—å£å·¦ä¸Šè§’åæ ‡
        p = ev->globalPos() - this->frameGeometry().topLeft();
    }
}

void MainWindow::mouseMoveEvent(QMouseEvent *ev)
{
    if (ev->buttons() & Qt::LeftButton && p.x() != 0 && p.y() != 0)
    {
        //ç§»åˆ°å·¦ä¸Šè§’
        move(ev->globalPos() - p);
    }
}

void MainWindow::mouseReleaseEvent(QMouseEvent *ev)
{
    p.setX(0);
    p.setY(0);
}

/*
 * pushMessageIntoChatWindowå‡½æ•°è¯´æ˜
 * bool typeï¼šå‘ç»™åˆ«äººæ¶ˆæ¯ï¼Œå³ä¾§æ°”æ³¡==trueï¼›æ¥æ”¶åˆ°æ¶ˆæ¯ï¼Œå·¦ä¾§æ°”æ³¡==falseã€‚
 * QImage msgï¼šå‘é€/æ¥æ”¶çš„Qimageã€‚
 * QString timeï¼šæ¶ˆæ¯æ—¶é—´ã€‚
 *      å‚è€ƒï¼štime = QString::number(QDateTime::currentDateTime().toTime_t());
 * bool isSending = falseï¼šæ˜¯å¦æ­£åœ¨å‘é€ï¼Œå¤§æ¦‚ç‡ç”¨ä¸åˆ°æ‰€ä»¥å…ˆè®¾ç½®æˆfalse
 *      ä»…å‘é€æ¶ˆæ¯æœ‰æ•ˆï¼Œæ¥æ”¶æ¶ˆæ¯æ—¶è¿™é‡Œå¯ä»¥éšä¾¿èµ‹å€¼ã€‚
 * QImage* image = NULL:å¤´åƒæŒ‡é’ˆï¼Œé»˜è®¤ä¸ºNULLï¼Œåˆ™æ˜¾ç¤ºé»˜è®¤å¤´åƒã€‚
 *
 */
void MainWindow::pushImageIntoChatWindow(bool type, QImage msg, QString time, QImage *image, bool isSending)
{
    if (type)
    {
        qDebug() << "me";
        if (isSending)
        {
            qDebug() << "sending";
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealImage(messageW, item, msg, time, ChatMessageWidget::Image_Me, image);
        }
        else
        {
            qDebug() << "not sending";
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealImage(messageW, item, msg, time, ChatMessageWidget::Image_Me, image);
            messageW->setTextSuccess();
        }
    }
    else
    {
        dealMessageTime(time);
        ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
        QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
        dealImage(messageW, item, msg, time, ChatMessageWidget::Image_She, image);
    }
    ui->listWidget->setCurrentRow(ui->listWidget->count() - 1);
}

void MainWindow::dealImage(ChatMessageWidget *messageW, QListWidgetItem *item, QImage img, QString time, ChatMessageWidget::User_Type type, QImage *image)
{
    messageW->setFixedWidth(ui->listWidget->width() - 25);
    int w = img.width(), h = img.height();
    QSize size = messageW->fontRect(QString("---------------\n---------------\n---------------\n---------------\n---------------\n---------------\n-"));
    qDebug() << size.width() << "...." << size.height();
    //    QSize size(300,150);
    item->setSizeHint(size);
    messageW->setImage(img, time, type, image);
    ui->listWidget->setItemWidget(item, messageW);
}
void MainWindow::changeMyIcon(QImage *uimg)
{
    if (uimg->isNull())
    {
        uimg = new QImage(":/assets/defaultHead.png");
    }
    QIcon qicon(QPixmap::fromImage(*uimg));
    ui->pushButton_image->setIcon(qicon);
}
void MainWindow::MoveFps() //çª—å£æŠ–åŠ¨ç‰¹æ•ˆ
{
    int befor_x = this->x();
    int befor_y = this->y();
    int move_x = 2;
    int move_y = 3;

    for (int i = 0; i < 10; i++)
    {
        this->move(befor_x + move_x, befor_y);
        Sleep(50);
        this->move(befor_x + move_x, befor_y + move_y);
        Sleep(50);
        this->move(befor_x, befor_y + move_y);
        Sleep(50);
        this->move(befor_x, befor_y);
        Sleep(50);
    }
}

void MainWindow::sendChatImage(const QString &imgFile, std::vector<string> userList, std::function<void()> onSuccess)
{
    createSendTask(
        udata, imgFile, [=](FileSock *skt, const QFileInfo &fileName, const QString &serverFileName)
        {
//        std::vector<std::string> userslist; // userListéœ€è¦ä»å¥½å‹æ å¯¼å…¥çš„ï¼Œè¿™é‡Œå›ºå®šçš„å ä¸ªä½ç½®
//        userslist.push_back(chattingInfo.chatFriend[0].userName);

        std::string data = Encoder_chatfile(serverFileName.toStdString(), udata.toStdString(), userList);
        QString packData = QString::fromStdString(data);
        client.sendMessage(packData);
        QImage chatimage(fileName.absoluteFilePath());
        pushImageIntoChatWindow(true, chatimage, QString::number(QDateTime::currentDateTime().toTime_t()), &user_image);
        onSuccess(); });
}


void MainWindow::pushFileIntoChatWindow(bool type,QString filePath, QString time, QImage *image, bool isSending)
{
    QString msg = filePath.right(filePath.lastIndexOf("/")+1);

    qDebug() << "addFile" << msg << time << ui->listWidget->count();
    if (type)
    {
        if (isSending)
        {
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealFile(messageW, item,filePath ,msg, time, ChatMessageWidget::File_Me, image);
        }
        else
        {
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealFile(messageW, item,filePath, msg, time, ChatMessageWidget::File_Me, image);
            messageW->setTextSuccess();
        }
    }
    else
    {
        dealMessageTime(time);
        ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
        QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
        dealFile(messageW, item,filePath, msg, time, ChatMessageWidget::User_She, image);
    }
    ui->listWidget->setCurrentRow(ui->listWidget->count() - 1);
}

void MainWindow::dealFile(ChatMessageWidget *messageW, QListWidgetItem *item,QString filePath, QString text, QString time, ChatMessageWidget::User_Type type, QImage *image)
{
    text = QString("ã€ğŸ“‚åŒå‡»æ‰“å¼€æ–‡ä»¶ã€‘")+text;
    messageW->setFixedWidth(ui->listWidget->width() - 25);
    QSize size = messageW->fontRect(text);
    item->setSizeHint(size);
    messageW->setFile(filePath,text, time, size, type, image);
    ui->listWidget->setItemWidget(item, messageW);
}

void MainWindow::changeIcon(QString username, QImage icon)
{
    friendList->changeIcon(username, icon);
}

void MainWindow::on_pushButton_3_clicked()
{
    QString filename = QFileDialog::getOpenFileName(this, "é€‰æ‹©æ–‡ä»¶");
    qDebug()<<filename;
    if(filename.length() ==0)
        return;
    QFileInfo fileinfo(filename);
    createSendTask(udata, filename, [=](FileSock *, const QFileInfo &, const QString &){
        vector<string> userList;
        for(auto v:chattingInfo.chatFriend)
            userList.push_back(v.userName);
        string packData = Encoder_chatfile(filename.toStdString(), udata.toStdString(),userList);
        client.sendMessage(QString::fromStdString(packData));
        pushFileIntoChatWindow(true, filename, QString::number(QDateTime::currentDateTime().toTime_t()), &user_image);
        msgRcd[QString::fromStdString(chattingInfo.chatFriend[0].userName)].push_back(MessageInfo(udata, filename, QString::number(QDateTime::currentDateTime().toTime_t()), MessageInfo::FILE));
    });
}

void MainWindow::on_pushButton_addgroup_clicked()
{
    group *g = new group(friendList, udata);
    g->show();
}
